exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -74,23 +83,35 @@ func NewPrwExporter(namespace string, endpoint string, client *http.Client, exte
 	}
 
 	userAgentHeader := fmt.Sprintf("%s/%s", strings.ReplaceAll(strings.ToLower(buildInfo.Description), " ", "-"), buildInfo.Version)
-
-	return &PrwExporter{
+	prwe := &PrwExporter{
 		namespace:       namespace,
 		externalLabels:  sanitizedLabels,
 		endpointURL:     endpointURL,
 		client:          client,
 		wg:              new(sync.WaitGroup),
 		closeChan:       make(chan struct{}),
+		walConfig:       walCfg,
 		userAgentHeader: userAgentHeader,
-	}, nil
+	}
+
+	if err := prwe.turnOnWALIfEnabled(); err != nil {
+		return nil, err
+	}
+	return prwe, nil
 }
 
+var errAlreadyClosed = errors.New("already closed")
+
 // Shutdown stops the exporter from accepting incoming calls(and return error), and wait for current export operations
 // to finish before returning
 func (prwe *PrwExporter) Shutdown(context.Context) error {
+	if prwe.closed {
+		return errAlreadyClosed
+	}
```

Comment:
It is guaranteed by the framework that Shutdown is called only once.

Reply:
I added this for extra protection because if someone accidentally invokes .Shutdown() in a defer and then later after processing an error, we'll have issues with the WAL and crashes too. This change brings even better protections, so I'd highly recommend we use the .closed guard.

Reply:
Shutdown is not deferred, and needs specific guarantees, see the other comment.

Reply:
@bogdandrutu can people import this code or not? If they can import specific components, they will misuse it. I honestly don't think making it more brittle is reasonable, yet preventing people from shooting their foot, but if that matches the review I'll change it.